glue/src/main/scala/com/gu/tableversions/glue/GlueMetastore.scala

```diff
@@ -0,0 +1,165 @@
+package com.gu.tableversions.glue
+
+import java.net.URI
+import cats.effect.Sync
+import cats.implicits._
+import com.amazonaws.services.glue.AWSGlue
+import com.amazonaws.services.glue.model.{Partition => GluePartition, Table => GlueTable, TableVersion => _, _}
+import com.gu.tableversions.core.Partition.{ColumnValue, PartitionColumn}
+import com.gu.tableversions.core._
+import com.gu.tableversions.metastore.Metastore.TableOperation
+import com.gu.tableversions.metastore.Metastore.TableOperation._
+import com.gu.tableversions.metastore.{Metastore, VersionPaths}
+
+import scala.collection.JavaConversions._
+
+class GlueMetastore[F[_]](glue: AWSGlue)(implicit F: Sync[F]) extends Metastore[F] {
+
+  override def currentVersion(table: TableName): F[TableVersion] = {
+
+    def getPartitionColumns(glueTable: GlueTable): List[PartitionColumn] =
+      Option(glueTable.getPartitionKeys).toList.flatten map { glueColumn =>
+        PartitionColumn(glueColumn.getName)
+      }
+
+    def snapshotTableVersion(tableLocation: URI): TableVersion =
+      SnapshotTableVersion(VersionPaths.parseVersion(tableLocation))
+
+    for {
+      glueTable <- getGlueTable(table)
+      partitionColumns <- F.delay(getPartitionColumns(glueTable))
+      tableLocation <- F.delay(findTableLocation(glueTable))
+      version <- if (partitionColumns.isEmpty)
+        F.delay(snapshotTableVersion(tableLocation))
+      else
+        getPartitionedTableVersion(table, tableLocation, partitionColumns)
+    } yield version
+  }
+
+  private def getPartitionedTableVersion(
+      table: TableName,
+      tableLocation: URI,
+      partitionColumns: List[PartitionColumn]): F[TableVersion] = {
+
+    def toPartition(columnValues: List[ColumnValue]): Partition = columnValues match {
+      case head :: tail => Partition(head, tail: _*)
+      case _            => throw new Exception("empty columnValues list for partition found")
```

Comment:
I think this is, like equivalent cases in the `SparkHiveMetastore`, cases that should never happen, i.e. invariants. Here for example, `getPartitionedTableVersion` should only be called for a partitioned table which will produce parittions with a valid set of colum values.

It would be nice if this can be narrowed down via types, for example making `partitionColumns` a `NonEmptyList` here, changing the way this is passed down. Don't know if this is really practical here though. If not, maybe the error in the `Exception` could be made clearer to indicate that this is an internal error? I.e. it could say "Invalid state: empty columnValues list for partition found", or "Internal error: ..." or something like that.

Reply:
columnValues is created combining 2 list returned by glue: one with the partition column names and the other with the corresponding values. 
I can make the partition column names one a nonEmpty list because the code is already explicitly checking in line 32 that there is at least one element. But the column names can still technically be empty even though it would make no sense for glue to return that.
The problem here is just made explicit by our definition of partition that requires a nonEmptyList so this error must be handled in some way.
In other parts of this repo we just threw exceptions in situations like this but we might want to refactor and make everything return proper types with room for errors if we think it's worth it.

Reply:
Good points. We probably should improve the testing of error scenarios like this and fix up the code to match. Not in this PR though!